File: server_python/services/auth_engine.py
# ==========================================
# AirOps AI — Auth Engine (Session Security)
# PII masking + secure PNR lookup
# ==========================================

import re
from db.sqlite_manager import get_db


def mask_sensitive_fields(data: dict) -> dict:
    """Mask PII fields in a dict recursively."""
    sensitive = ["cpf", "document_number", "email", "phone", "credit_card"]
    masked = dict(data)

    for key in list(masked.keys()):
        val = str(masked.get(key, "") or "")
        k = key.lower()

        if any(f in k for f in sensitive):
            if "cpf" in k or "document" in k:
                masked[key] = f"***{val[-4:]}" if len(val) > 4 else "***"
            elif "email" in k:
                masked[key] = re.sub(r"(.{2})(.*)(@.*)", r"\1***\3", val) if "@" in val else "***"
            elif "phone" in k:
                masked[key] = f"***{val[-4:]}" if len(val) > 4 else "***"
            elif "credit" in k:
                masked[key] = f"****-****-****-{val[-4:]}" if len(val) > 4 else "****"

        if isinstance(masked[key], dict):
            masked[key] = mask_sensitive_fields(masked[key])

    return masked


def secure_pnr_lookup(locator: str, last_name: str | None = None) -> dict:
    """Lookup PNR with security scope, joining all related tables."""
    db = get_db()

    pnr = db.execute("SELECT * FROM pnr_reservations WHERE locator = ?", (locator,)).fetchone()
    if not pnr:
        return {"found": False, "error": "PNR não encontrado."}

    passengers = [dict(r) for r in db.execute("SELECT * FROM passengers WHERE pnr = ?", (locator,)).fetchall()]

    # Guardrail: If last_name provided, validate
    if last_name:
        match = any(p["last_name"].lower() == last_name.lower() for p in passengers)
        if not match:
            return {"found": False, "error": "Sobrenome não corresponde à reserva."}

    raw_segments = [dict(r) for r in db.execute("""
        SELECT fs.* FROM flight_segments fs
        JOIN pnr_segments ps ON ps.segment_id = fs.id
        WHERE ps.pnr = ?
        ORDER BY ps.sequence_number
    """, (locator,)).fetchall()]
    
    segments = []
    for s in raw_segments:
        segments.append({
            "id": s.get("id"),
            "flightNumber": s.get("flight_number"),
            "airline": s.get("airline"),
            "origin": {"code": s.get("origin")},
            "destination": {"code": s.get("destination")},
            "scheduledDeparture": s.get("scheduled_departure"),
            "actualDeparture": s.get("actual_departure"),
            "scheduledArrival": s.get("scheduled_arrival"),
            "status": s.get("segment_status"),
            "delayMinutes": s.get("delay_minutes"),
            "gate": s.get("gate"),
            "aircraft": s.get("aircraft"),
            "seatNumber": s.get("seat")
        })

    raw_tickets = [dict(r) for r in db.execute("""
        SELECT t.*, fr.fare_family, fr.refundable, fr.change_allowed, fr.change_fee,
               fr.cancel_fee, fr.voucher_allowed, fr.free_rebooking_on_irop,
               fr.baggage_included, fr.baggage_weight_kg, fr.description as fare_description
        FROM tickets t
        LEFT JOIN fare_rules fr ON t.fare_basis = fr.fare_basis
        WHERE t.pnr = ?
    """, (locator,)).fetchall()]
    
    tickets = []
    for t in raw_tickets:
        tickets.append({
            "eTicket": t.get("ticket_number"),
            "fareBasis": t.get("fare_basis"),
            "amount": t.get("total_amount") or 0,
            "fareRules": {
                "refundable": bool(t.get("refundable")),
                "changeAllowed": bool(t.get("change_allowed")),
                "changeFee": t.get("change_fee") or 0,
                "cancelFee": t.get("cancel_fee") or 0,
                "creditEligible": bool(t.get("voucher_allowed"))
            }
        })

    baggage = [dict(r) for r in db.execute("SELECT * FROM baggage_items WHERE pnr = ?", (locator,)).fetchall()]

    loyalty = None
    if passengers and passengers[0].get("customer_id"):
        row = db.execute("SELECT * FROM loyalty_profiles WHERE customer_id = ?", (passengers[0]["customer_id"],)).fetchone()
        if row:
            loyalty = dict(row)
            
    mapped_passengers = []
    for p in passengers:
        mapped_passengers.append({
            "id": p.get("id"),
            "firstName": p.get("first_name"),
            "lastName": p.get("last_name"),
            "email": p.get("email"),
            "phone": p.get("phone"),
            "loyaltyTier": loyalty.get("tier") if loyalty else "standard",
            "loyaltyMiles": loyalty.get("miles_balance") if loyalty else 0,
        })
            
    # Guardrail 3: Never expose full details to agent context, ONLY masked
    masked_pnr = dict(pnr)
    masked_pnr = mask_sensitive_fields(masked_pnr)

    return {
        "found": True,
        "pnr": {
            "locator": locator,
            **masked_pnr,
            "passengers": [mask_sensitive_fields(p) for p in mapped_passengers],
            "segments": segments,
            "tickets": tickets,
            "baggage": baggage,
            "loyalty": loyalty,
        },
    }

File: server/src/services/authEngine.ts
// ==========================================
// AirOps AI — Auth Engine (Session Security)
// Ensures data isolation per customer/session
// ==========================================

import { db } from '../db/sqlite.js';

export type AuthLevel = 'none' | 'pnr_only' | 'pnr_lastname' | 'pnr_lastname_doc' | 'otp_verified';

export interface SessionAuth {
  sessionId: string;
  customerId: string | null;
  pnr: string | null;
  authLevel: AuthLevel;
  authenticated: boolean;
}

export interface ActionAuth {
  action: string;
  requiredLevel: AuthLevel;
  requiresCustomerMatch: boolean;
  requiresPnrMatch: boolean;
}

// ---- Action → Minimum Auth Level ----
const ACTION_AUTH_MAP: Record<string, ActionAuth> = {
  get_flight_status: { action: 'get_flight_status', requiredLevel: 'pnr_only', requiresCustomerMatch: false, requiresPnrMatch: true },
  lookup_pnr: { action: 'lookup_pnr', requiredLevel: 'pnr_lastname', requiresCustomerMatch: false, requiresPnrMatch: true },
  get_passenger_details: { action: 'get_passenger_details', requiredLevel: 'pnr_lastname', requiresCustomerMatch: true, requiresPnrMatch: true },
  get_anac_rules: { action: 'get_anac_rules', requiredLevel: 'none', requiresCustomerMatch: false, requiresPnrMatch: false },
  check_refund_eligibility: { action: 'check_refund_eligibility', requiredLevel: 'pnr_lastname', requiresCustomerMatch: true, requiresPnrMatch: true },
  execute_refund: { action: 'execute_refund', requiredLevel: 'pnr_lastname_doc', requiresCustomerMatch: true, requiresPnrMatch: true },
  execute_rebook: { action: 'execute_rebook', requiredLevel: 'pnr_lastname', requiresCustomerMatch: true, requiresPnrMatch: true },
  issue_voucher: { action: 'issue_voucher', requiredLevel: 'pnr_lastname', requiresCustomerMatch: true, requiresPnrMatch: true },
  transfer_to_human: { action: 'transfer_to_human', requiredLevel: 'none', requiresCustomerMatch: false, requiresPnrMatch: false },
};

// ---- Validate that session can access a PNR ----

export function validatePnrAccess(sessionAuth: SessionAuth, requestedPnr: string): { allowed: boolean; reason: string } {
  // If no PNR set in session yet, allow first lookup
  if (!sessionAuth.pnr) {
    return { allowed: true, reason: 'First PNR lookup in session' };
  }

  // PNR must match session
  if (sessionAuth.pnr !== requestedPnr) {
    return {
      allowed: false,
      reason: 'Não consegui validar a titularidade dessa reserva com os dados informados. Para sua segurança, vou te encaminhar para um atendente.',
    };
  }

  return { allowed: true, reason: 'PNR matches session' };
}

// ---- Check if action is authorized ----

export function isActionAuthorized(sessionAuth: SessionAuth, action: string): { allowed: boolean; reason: string } {
  const authReq = ACTION_AUTH_MAP[action];
  if (!authReq) {
    return { allowed: false, reason: `Action "${action}" not recognized` };
  }

  // Check auth level hierarchy
  const levels: AuthLevel[] = ['none', 'pnr_only', 'pnr_lastname', 'pnr_lastname_doc', 'otp_verified'];
  const sessionLevel = levels.indexOf(sessionAuth.authLevel);
  const requiredLevel = levels.indexOf(authReq.requiredLevel);

  if (sessionLevel < requiredLevel) {
    return {
      allowed: false,
      reason: `Ação "${action}" requer autenticação nível "${authReq.requiredLevel}". Nível atual: "${sessionAuth.authLevel}".`,
    };
  }

  return { allowed: true, reason: 'Authorized' };
}

// ---- Mask sensitive data in output ----

export function maskSensitiveFields(data: Record<string, any>): Record<string, any> {
  const masked = { ...data };
  const sensitiveFields = ['cpf', 'document_number', 'email', 'phone', 'credit_card'];

  for (const key of Object.keys(masked)) {
    if (sensitiveFields.some(f => key.toLowerCase().includes(f))) {
      const val = String(masked[key] || '');
      if (key.toLowerCase().includes('cpf') || key.toLowerCase().includes('document')) {
        masked[key] = val.length > 4 ? `***${val.slice(-4)}` : '***';
      } else if (key.toLowerCase().includes('email')) {
        masked[key] = val.replace(/(.{2})(.*)(@.*)/, '$1***$3');
      } else if (key.toLowerCase().includes('phone')) {
        masked[key] = val.length > 4 ? `***${val.slice(-4)}` : '***';
      } else if (key.toLowerCase().includes('credit')) {
        masked[key] = val.length > 4 ? `****-****-****-${val.slice(-4)}` : '****';
      }
    }

    // Recurse into nested objects
    if (masked[key] && typeof masked[key] === 'object' && !Array.isArray(masked[key])) {
      masked[key] = maskSensitiveFields(masked[key]);
    }
  }

  return masked;
}

// ---- Lookup PNR with security scope ----

export function securePnrLookup(locator: string, lastName?: string) {
  const pnr = db.prepare('SELECT * FROM pnr_reservations WHERE locator = ?').get(locator) as any;
  if (!pnr) return { found: false, error: 'PNR não encontrado.' };

  const passengers = db.prepare('SELECT * FROM passengers WHERE pnr = ?').all(locator) as any[];

  // If lastName provided, validate
  if (lastName) {
    const match = passengers.some(p => p.last_name.toLowerCase() === lastName.toLowerCase());
    if (!match) {
      return { found: false, error: 'Sobrenome não corresponde à reserva.' };
    }
  }

  // Get linked flight segments
  const segments = db.prepare(`
    SELECT fs.* FROM flight_segments fs
    JOIN pnr_segments ps ON ps.segment_id = fs.id
    WHERE ps.pnr = ?
    ORDER BY ps.sequence_number
  `).all(locator) as any[];

  // Get tickets with fare rules
  const tickets = db.prepare(`
    SELECT t.*, fr.fare_family, fr.refundable, fr.change_allowed, fr.change_fee,
           fr.cancel_fee, fr.voucher_allowed, fr.free_rebooking_on_irop,
           fr.baggage_included, fr.baggage_weight_kg, fr.description as fare_description
    FROM tickets t
    LEFT JOIN fare_rules fr ON t.fare_basis = fr.fare_basis
    WHERE t.pnr = ?
  `).all(locator) as any[];

  // Get baggage
  const baggage = db.prepare('SELECT * FROM baggage_items WHERE pnr = ?').all(locator) as any[];

  // Get loyalty for primary passenger
  const primaryPax = passengers[0];
  let loyalty = null;
  if (primaryPax?.customer_id) {
    loyalty = db.prepare('SELECT * FROM loyalty_profiles WHERE customer_id = ?').get(primaryPax.customer_id);
  }

  return {
    found: true,
    pnr: {
      ...pnr,
      passengers: passengers.map(p => maskSensitiveFields(p)),
      segments,
      tickets,
      baggage,
      loyalty,
    },
  };
}
